README.md

```diff
@@ -2,7 +2,157 @@
+let result = ASON.deserialize<StaticArray<u8>>(buffer);
+
+assert(result.length == 3);
+assert(result[0] == <u8>29);
+assert(result[0] == <u8>6);
+assert(result[0] == <u8>4);
+```
+
+It's also possible to save heap allocations and re-use a `Serializer` and `Deserializer` object when serializing multiple objects of the same type.
+
+```ts
+import { Serializer, Deserializer } from "@ason/assembly";
+class Vec3 {
+  constructor(
+    public x: f32,
+    public y: f32,
+    public z: f32,
+  ) {}
+}
+
+let result = new Array<StaticArray<u8>>(); // an array of buffers
+let serializer = new ASON.Serializer<Vec3>();
+
+for (let i = 0; i < 10; i++) {
+  result.push(serializer.serialize(new Vec3(1, 2, 3)));
+}
+
+let deserializer = new ASON.Deserializer<Vec3>();
+
+for (let i = 0; i < 10; i++) {
+  let vec = deserializer.deserialize(result[i]);
+  assert(vec); // make sure the reference isn't null
+  assert(vec.x == 1); // check the properties
+  assert(vec.y == 2);
+  assert(vec.x == 3);
+}
+```
+
+# Implementation
+
+In order to assemble an AssemblyScript object, we need a few key pieces of data about our reference.
+
+```ts
+@unmanaged export class ReferenceEntry {
+  rttid: u32; // The type Id
+  entryId: u32; // some kind of unique entry identifier
+  offset: usize; // and how big it is
+}
+```
+
+We now have enough information to make a list of references. Think of a c-like `ReferenceEntry*` pointer. For every reference in an object tree, we can allocate some space in a buffer to describe it's shape. The first step is to visit each reference on the object tree, and then allocate a `ReferenceEntry` for it. Then, we allocate a `LinkEntry` that describes the relationship between the parent and the child references.
+
+```ts
+@unmanaged
+export class LinkEntry {
+  parentEntryId: u32; // The parent entry id
+  offset: usize; // Where it should be stored on the parent
+  childEntryId: u32; // The child entry id
+}
+```
+
+Finally, we need to store the fields that are numeric. We can allocate a `FieldEntry` on another table to describe each field on each reference.
+
+```ts
+
+@unmanaged
+export class FieldEntry8 {
+  entryId: u32;
+  offset: usize;
+  value: u8;
+}
+
+@unmanaged
+export class FieldEntry16 {
+  entryId: u32;
+  offset: usize;
+  value: u16;
+}
+
+@unmanaged
+export class FieldEntry32 {
+  entryId: u32;
+  offset: usize;
+  value: u32;
+}
+
+@unmanaged
+export class FieldEntry64 {
+  entryId: u32;
+  offset: usize;
+  value: u64;
+}
+```
+
+Since there are only four different field sizes, it makes sense to describe an object using four different field types. Storing the numeric type is unnecessary, because the bytes can be stored and restored, as long as size is properly accounted for without the need to cast numeric values to floats or integers.
+
+Lastly, we assert that the `entryId` stored at entry `0` is the primary entry a buffer. This of course assumes the generic type of the `Serializer` used is actually a reference.
+
```

Comment:
Can you add a complete example here? i.e one that shows how the above information is used by `ASON.Serializer`?

Reply:
I'm thinking that maybe this information isn't necessary here. We should glorify ASON as an airtight box that does a lot of magic. :)